images/kube-webhook-certgen/rootfs/pkg/k8s/k8s.go

```diff
@@ -2,110 +2,222 @@ package k8s
+	aggregatorClientset clientset.Interface
 }
 
-func New(clientset kubernetes.Interface) *k8s {
+func New(clientset kubernetes.Interface, aggregatorClientset clientset.Interface) *k8s {
 	if clientset == nil {
 		log.Fatal("no kubernetes client given")
 	}
 
+	if aggregatorClientset == nil {
+		log.Fatal("no kubernetes aggregator client given")
+	}
+
 	return &k8s{
-		clientset: clientset,
+		clientset:           clientset,
+		aggregatorClientset: aggregatorClientset,
 	}
 }
 
+type PatchOptions struct {
+	ValidatingWebhookConfigurationName string
+	MutatingWebhookConfigurationName   string
+	APIServiceName                     string
+	CABundle                           []byte
+	FailurePolicyType                  admissionv1.FailurePolicyType
+}
+
+func (k8s *k8s) PatchObjects(ctx context.Context, options PatchOptions) error {
+	patchMutating := options.MutatingWebhookConfigurationName != ""
+	patchValidating := options.ValidatingWebhookConfigurationName != ""
+	patchAPIService := options.APIServiceName != ""
+
+	if !patchMutating && !patchValidating && options.FailurePolicyType != "" {
+		return fmt.Errorf("failurePolicy specified, but no webhook will be patched")
+	}
+
+	if options.MutatingWebhookConfigurationName != options.ValidatingWebhookConfigurationName {
+		return fmt.Errorf("webhook names must be the same")
+	}
+
+	if patchAPIService {
+		log.Infof("patching APIService %q", options.APIServiceName)
+
+		if err := k8s.patchAPIService(ctx, options.APIServiceName, options.CABundle); err != nil {
+			// Intentionally don't wrap error here to preserve old behavior and be able to log both
+			// original error and a message.
+			return err
+		}
+	}
+
+	if patchMutating || patchValidating {
+		return k8s.patchWebhookConfigurations(ctx, options.ValidatingWebhookConfigurationName, options.CABundle, options.FailurePolicyType, patchMutating, patchValidating)
+	}
+
+	return nil
+}
+
+func (k8s *k8s) patchAPIService(ctx context.Context, objectName string, ca []byte) error {
+	log.Infof("patching APIService %q", objectName)
+
+	c := k8s.aggregatorClientset.ApiregistrationV1().APIServices()
+
+	apiService, err := c.Get(ctx, objectName, metav1.GetOptions{})
+	if err != nil {
+		return &wrappedError{
+			err:     err,
+			message: fmt.Sprintf("failed getting APIService %q", objectName),
+		}
+	}
+
+	apiService.Spec.CABundle = ca
+	apiService.Spec.InsecureSkipTLSVerify = false
+
+	if _, err := c.Update(ctx, apiService, metav1.UpdateOptions{}); err != nil {
+		return &wrappedError{
+			err:     err,
+			message: fmt.Sprintf("failed patching APIService %q", objectName),
+		}
+	}
+
+	log.Debug("patched APIService")
+
+	return nil
+}
+
 // PatchWebhookConfigurations will patch validatingWebhook and mutatingWebhook clientConfig configurations with
 // the provided ca data. If failurePolicy is provided, patch all webhooks with this value
-func (k8s *k8s) PatchWebhookConfigurations(
+func (k8s *k8s) patchWebhookConfigurations(
```

Comment:
nit: As you are unexporting the function (still wondering why!) you should change the comment as well

Reply:
>still wondering why

To reduce the surface of the package and to promote use of newly introduced and more flexible `PatchObjects` function. Though this is a breaking change on Go module level, so I can revert this change if you like.

> you should change the comment as well

Probably, but then some of my linters yell, that comment is not capitalized, so I recently just started leaving those docs as they are.

Please suggest a desired change and I'll apply it, as I don't have a preference here.